<!DOCTYPE html><html lang="zh" data-theme="light"><head><meta charset="UTF-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0,viewport-fit=cover"><title>如何使用TCGAbiolinks下载TCGA数据并整理 | Yezi's Hexo Blog</title><meta name="author" content="Yezi"><meta name="copyright" content="Yezi"><meta name="format-detection" content="telephone=no"><meta name="theme-color" content="ffffff"><meta name="description" content="引言一般来讲，我们想要使用TCGA数据，大概有三种方法，一是直接从GDC官网或官方下载工具gdc-client下载文件后自行处理，二是使用数据库如UCSC Xena或Firehouse，三是使用TCGAbiolinks R包自动下载并处理。 从官网下载并不麻烦，但是第一是需要选取非常多的自定义选项，第二是网络环境不好会容易中断，对于初学者倒是一个非常好的了解生物信息学的途径，但遇到批量化处理需求的">
<meta property="og:type" content="article">
<meta property="og:title" content="如何使用TCGAbiolinks下载TCGA数据并整理">
<meta property="og:url" content="https://yeyeziblog.eu.org/2023/03/23/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/TCGA/TCGAbiolinks%E5%8C%85%E5%8F%82%E6%95%B0%E8%AF%A6%E8%A7%A3/index.html">
<meta property="og:site_name" content="Yezi&#39;s Hexo Blog">
<meta property="og:description" content="引言一般来讲，我们想要使用TCGA数据，大概有三种方法，一是直接从GDC官网或官方下载工具gdc-client下载文件后自行处理，二是使用数据库如UCSC Xena或Firehouse，三是使用TCGAbiolinks R包自动下载并处理。 从官网下载并不麻烦，但是第一是需要选取非常多的自定义选项，第二是网络环境不好会容易中断，对于初学者倒是一个非常好的了解生物信息学的途径，但遇到批量化处理需求的">
<meta property="og:locale">
<meta property="og:image" content="https://yeyeziblog.eu.org/img/TCGA%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD_2023%E5%B9%B403%E6%9C%88.jpg">
<meta property="article:published_time" content="2023-03-23T12:26:48.271Z">
<meta property="article:modified_time" content="2023-03-23T12:26:48.271Z">
<meta property="article:author" content="Yezi">
<meta property="article:tag" content="R 语言">
<meta property="article:tag" content="TCGA">
<meta name="twitter:card" content="summary">
<meta name="twitter:image" content="https://yeyeziblog.eu.org/img/TCGA%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD_2023%E5%B9%B403%E6%9C%88.jpg"><link rel="shortcut icon" href="/img/favicon.png"><link rel="canonical" href="https://yeyeziblog.eu.org/2023/03/23/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/TCGA/TCGAbiolinks%E5%8C%85%E5%8F%82%E6%95%B0%E8%AF%A6%E8%A7%A3/index.html"><link rel="preconnect" href="//cdn.jsdelivr.net"/><link rel="preconnect" href="//busuanzi.ibruce.info"/><link rel="stylesheet" href="/css/index.css"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fortawesome/fontawesome-free/css/all.min.css" media="print" onload="this.media='all'"><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox/fancybox.min.css" media="print" onload="this.media='all'"><script>const GLOBAL_CONFIG = { 
  root: '/',
  algolia: {"appId":"74HMPRFIN5","apiKey":"896979d16b9d8fcb17ef98c222d6d40a","indexName":"yeziblog","hits":{"per_page":10},"languages":{"input_placeholder":"Search for Posts","hits_empty":"We didn't find any results for the search: ${query}.","hits_stats":"${hits} results found in ${time} ms"}},
  localSearch: undefined,
  translate: undefined,
  noticeOutdate: undefined,
  highlight: {"plugin":"highlighjs","highlightCopy":true,"highlightLang":true,"highlightHeightLimit":false},
  copy: {
    success: 'Copy successfully',
    error: 'Copy error',
    noSupport: 'The browser does not support'
  },
  relativeDate: {
    homepage: false,
    post: false
  },
  runtime: '',
  dateSuffix: {
    just: 'Just',
    min: 'minutes ago',
    hour: 'hours ago',
    day: 'days ago',
    month: 'months ago'
  },
  copyright: undefined,
  lightbox: 'fancybox',
  Snackbar: undefined,
  source: {
    justifiedGallery: {
      js: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.js',
      css: 'https://cdn.jsdelivr.net/npm/flickr-justified-gallery/dist/fjGallery.min.css'
    }
  },
  isPhotoFigcaption: false,
  islazyload: false,
  isAnchor: false,
  percent: {
    toc: true,
    rightside: false,
  }
}</script><script id="config-diff">var GLOBAL_CONFIG_SITE = {
  title: '如何使用TCGAbiolinks下载TCGA数据并整理',
  isPost: true,
  isHome: false,
  isHighlightShrink: false,
  isToc: true,
  postUpdate: '2023-03-23 20:26:48'
}</script><noscript><style type="text/css">
  #nav {
    opacity: 1
  }
  .justified-gallery img {
    opacity: 1
  }

  #recent-posts time,
  #post-meta time {
    display: inline !important
  }
</style></noscript><script>(win=>{
    win.saveToLocal = {
      set: function setWithExpiry(key, value, ttl) {
        if (ttl === 0) return
        const now = new Date()
        const expiryDay = ttl * 86400000
        const item = {
          value: value,
          expiry: now.getTime() + expiryDay,
        }
        localStorage.setItem(key, JSON.stringify(item))
      },

      get: function getWithExpiry(key) {
        const itemStr = localStorage.getItem(key)

        if (!itemStr) {
          return undefined
        }
        const item = JSON.parse(itemStr)
        const now = new Date()

        if (now.getTime() > item.expiry) {
          localStorage.removeItem(key)
          return undefined
        }
        return item.value
      }
    }
  
    win.getScript = url => new Promise((resolve, reject) => {
      const script = document.createElement('script')
      script.src = url
      script.async = true
      script.onerror = reject
      script.onload = script.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        script.onload = script.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(script)
    })
  
    win.getCSS = (url,id = false) => new Promise((resolve, reject) => {
      const link = document.createElement('link')
      link.rel = 'stylesheet'
      link.href = url
      if (id) link.id = id
      link.onerror = reject
      link.onload = link.onreadystatechange = function() {
        const loadState = this.readyState
        if (loadState && loadState !== 'loaded' && loadState !== 'complete') return
        link.onload = link.onreadystatechange = null
        resolve()
      }
      document.head.appendChild(link)
    })
  
      win.activateDarkMode = function () {
        document.documentElement.setAttribute('data-theme', 'dark')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', '#0d0d0d')
        }
      }
      win.activateLightMode = function () {
        document.documentElement.setAttribute('data-theme', 'light')
        if (document.querySelector('meta[name="theme-color"]') !== null) {
          document.querySelector('meta[name="theme-color"]').setAttribute('content', 'ffffff')
        }
      }
      const t = saveToLocal.get('theme')
    
          if (t === 'dark') activateDarkMode()
          else if (t === 'light') activateLightMode()
        
      const asideStatus = saveToLocal.get('aside-status')
      if (asideStatus !== undefined) {
        if (asideStatus === 'hide') {
          document.documentElement.classList.add('hide-aside')
        } else {
          document.documentElement.classList.remove('hide-aside')
        }
      }
    
    const detectApple = () => {
      if(/iPad|iPhone|iPod|Macintosh/.test(navigator.userAgent)){
        document.documentElement.classList.add('apple')
      }
    }
    detectApple()
    })(window)</script><meta name="generator" content="Hexo 6.3.0"></head><body><div id="web_bg"></div><div id="sidebar"><div id="menu-mask"></div><div id="sidebar-menus"><div class="avatar-img is-center"><img src="/image/avatar.png" onerror="onerror=null;src='/img/friend_404.gif'" alt="avatar"/></div><div class="sidebar-site-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">152</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">91</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">42</div></a></div><hr/><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div></div></div><div class="post" id="body-wrap"><header class="not-top-img" id="page-header"><nav id="nav"><span id="blog-info"><a href="/" title="Yezi's Hexo Blog"><span class="site-name">Yezi's Hexo Blog</span></a></span><div id="menus"><div id="search-button"><a class="site-page social-icon search" href="javascript:void(0);"><i class="fas fa-search fa-fw"></i><span> Search</span></a></div><div class="menus_items"><div class="menus_item"><a class="site-page" href="/"><i class="fa-fw fas fa-home"></i><span> Home</span></a></div><div class="menus_item"><a class="site-page" href="/archives/"><i class="fa-fw fas fa-archive"></i><span> Archives</span></a></div><div class="menus_item"><a class="site-page" href="/tags/"><i class="fa-fw fas fa-tags"></i><span> Tags</span></a></div><div class="menus_item"><a class="site-page" href="/categories/"><i class="fa-fw fas fa-folder-open"></i><span> Categories</span></a></div><div class="menus_item"><a class="site-page" href="/link/"><i class="fa-fw fas fa-link"></i><span> Link</span></a></div><div class="menus_item"><a class="site-page" href="/about/"><i class="fa-fw fas fa-heart"></i><span> About</span></a></div></div><div id="toggle-menu"><a class="site-page" href="javascript:void(0);"><i class="fas fa-bars fa-fw"></i></a></div></div></nav></header><main class="layout" id="content-inner"><div id="post"><div id="post-info"><h1 class="post-title">如何使用TCGAbiolinks下载TCGA数据并整理</h1><div id="post-meta"><div class="meta-firstline"><span class="post-meta-date"><i class="far fa-calendar-alt fa-fw post-meta-icon"></i><span class="post-meta-label">Created</span><time class="post-meta-date-created" datetime="2023-03-23T12:26:48.271Z" title="Created 2023-03-23 20:26:48">2023-03-23</time><span class="post-meta-separator">|</span><i class="fas fa-history fa-fw post-meta-icon"></i><span class="post-meta-label">Updated</span><time class="post-meta-date-updated" datetime="2023-03-23T12:26:48.271Z" title="Updated 2023-03-23 20:26:48">2023-03-23</time></span><span class="post-meta-categories"><span class="post-meta-separator">|</span><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/">生物信息学</a><i class="fas fa-angle-right post-meta-separator"></i><i class="fas fa-inbox fa-fw post-meta-icon"></i><a class="post-meta-categories" href="/categories/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/TCGA/">TCGA</a></span></div><div class="meta-secondline"><span class="post-meta-separator">|</span><span class="post-meta-pv-cv" id="" data-flag-title="如何使用TCGAbiolinks下载TCGA数据并整理"><i class="far fa-eye fa-fw post-meta-icon"></i><span class="post-meta-label">Post View:</span><span id="busuanzi_value_page_pv"><i class="fa-solid fa-spinner fa-spin"></i></span></span></div></div></div><article class="post-content" id="article-container"><h2 id="引言"><a href="#引言" class="headerlink" title="引言"></a>引言</h2><p>一般来讲，我们想要使用TCGA数据，大概有三种方法，一是直接从<a target="_blank" rel="noopener" href="https://portal.gdc.cancer.gov/">GDC官网</a>或<a target="_blank" rel="noopener" href="https://docs.gdc.cancer.gov/Data_Transfer_Tool/Users_Guide/Getting_Started/">官方下载工具gdc-client</a>下载文件后自行处理，二是使用数据库如<a target="_blank" rel="noopener" href="https://xena.ucsc.edu/">UCSC Xena</a>或<a target="_blank" rel="noopener" href="https://gdac.broadinstitute.org/">Firehouse</a>，三是使用<a target="_blank" rel="noopener" href="https://bioconductor.org/packages/release/bioc/html/TCGAbiolinks.html">TCGAbiolinks</a> R包自动下载并处理。</p>
<p>从官网下载并不麻烦，但是第一是需要选取非常多的自定义选项，第二是网络环境不好会容易中断，对于初学者倒是一个非常好的了解生物信息学的途径，但遇到批量化处理需求的时候就会难以进行。其后是数据库法，数据库虽然方便，但是并不会随着官网的更新变动，如 <code>GDC Xena Hub</code> 最后一次更新时间是 <code>2019-08-28</code> ， <code>Firehose</code> 更是停留在了更遥远的 <code>2016_01_28</code> ……</p>
<p>那么, 如果我需要批量下载的话, 难道我需要一个个的从网页加入Cart获取mata吗, 我不要……</p>
<p>幸好，已经有人造了非常好用的轮子，当然可以轻松学习一下用起来啦。</p>
<p><code>TCGAbiolinks</code> 包是从TCGA数据库官网接口下载数据的R包。它的一些函数能够轻松地帮我们下载数据和整理数据格式。其实就是broad研究所的firehose命令行工具的R包装！</p>
<p>需要注意的是，2022年TCGA数据库进行了一次比较大的更新，其中包括了数据格式的变动，因此 <code>TCGAbiolinks</code> 也必须随之更新到最新版。下面，正式开始。</p>
<h2 id="效果展示"><a href="#效果展示" class="headerlink" title="效果展示"></a>效果展示</h2><p>可获得文件如下:</p>
<ol>
<li>TCGA转录组数据原始文件(tsv)及临床原始文件(xml), 均附带清单</li>
<li>表达矩阵表格(可选<code>&quot;counts&quot;, &quot;fpkm&quot;, &quot;tpm&quot;</code>)</li>
<li>分组文件</li>
<li>临床数据, 其中包含生存数据</li>
</ol>
<p><img src="/img/20230324142741.png" alt="可获得文件展示"></p>
<p><img src="/img/20230324143537.png" alt="counts值表达矩阵"></p>
<p><img src="/img/20230324143614.png" alt="分组文件"></p>
<p><img src="/img/20230324143028.png" alt="Phenodata.csv"></p>
<p><img src="/img/20230324142931.png" alt="Survivalata.csv"></p>
<h2 id="过程"><a href="#过程" class="headerlink" title="过程"></a>过程</h2><h3 id="下载"><a href="#下载" class="headerlink" title="下载"></a>下载</h3><p>首先是更新最新版的 <code>TCGAbiolinks</code> 包, 我使用的办法是使用Clash获得本地代理后对 <code>R session</code> 进行代理流量转发, 而后直接运行 <code>BiocManager::install(&quot;TCGAbiolinks&quot;)</code> , 选择所有包更新, 很顺利的就更新好了.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br></pre></td><td class="code"><pre><span class="line">Sys.setenv<span class="punctuation">(</span>http_proxy<span class="operator">=</span><span class="string">&quot;http://127.0.0.1:7890&quot;</span><span class="punctuation">)</span></span><br><span class="line">Sys.setenv<span class="punctuation">(</span>https_proxy<span class="operator">=</span><span class="string">&quot;http://127.0.0.1:7890&quot;</span><span class="punctuation">)</span></span><br><span class="line">Sys.setenv<span class="punctuation">(</span>all_proxy<span class="operator">=</span><span class="string">&quot;socks5://127.0.0.1:7890&quot;</span><span class="punctuation">)</span></span><br><span class="line">BiocManager<span class="operator">::</span>install<span class="punctuation">(</span><span class="string">&quot;TCGAbiolinks&quot;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>其中, <code>127.0.0.1</code> 是本机(local_host)的ip地址, 而7890是Clash的默认端口.</p>
<p>安装成功后，就可以开始使用了。</p>
<p>如, 运行 <code>TCGAbiolinks:::getGDCprojects()$project_id</code> 获取各个癌种的项目id, 总计有74个.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line">library<span class="punctuation">(</span>TCGAbiolinks<span class="punctuation">)</span></span><br><span class="line">TCGAbiolinks<span class="operator">:::</span>getGDCprojects<span class="punctuation">(</span><span class="punctuation">)</span><span class="operator">$</span>project_id</span><br><span class="line"><span class="comment"># [1] &quot;GENIE-GRCC&quot;                &quot;GENIE-DFCI&quot;                &quot;GENIE-NKI&quot;                 &quot;GENIE-VICC&quot;               </span></span><br><span class="line"><span class="comment"># [5] &quot;GENIE-UHN&quot;                 &quot;GENIE-MDA&quot;                 &quot;GENIE-MSK&quot;                 &quot;GENIE-JHU&quot;    </span></span><br><span class="line"><span class="comment">#  ...</span></span><br><span class="line"> TCGAbiolinks<span class="operator">:::</span>getGDCprojects<span class="punctuation">(</span><span class="punctuation">)</span><span class="operator">$</span>project_id <span class="operator">%&gt;%</span> <span class="built_in">length</span><span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># [1] 74</span></span><br></pre></td></tr></table></figure>

<p>如需获取TCGA癌症数据, 可以使用正则表达式获取开头带有 <code>TCGA</code> 的项目.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">projects <span class="operator">&lt;-</span> TCGAbiolinks<span class="operator">::</span>getGDCprojects<span class="punctuation">(</span><span class="punctuation">)</span><span class="operator">$</span>project_id <span class="comment">## 获取癌症名字</span></span><br><span class="line">projects <span class="operator">&lt;-</span> projects<span class="punctuation">[</span>grepl<span class="punctuation">(</span><span class="string">&quot;^TCGA&quot;</span><span class="punctuation">,</span> projects<span class="punctuation">,</span> perl <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">projects</span><br><span class="line"><span class="comment">#  [1] &quot;TCGA-CHOL&quot; &quot;TCGA-LIHC&quot; &quot;TCGA-DLBC&quot; &quot;TCGA-BLCA&quot; &quot;TCGA-ACC&quot;  &quot;TCGA-CESC&quot; &quot;TCGA-PCPG&quot; &quot;TCGA-PAAD&quot; &quot;TCGA-MESO&quot; &quot;TCGA-TGCT&quot;</span></span><br><span class="line"><span class="comment"># [11] &quot;TCGA-KIRP&quot; &quot;TCGA-UVM&quot;  &quot;TCGA-UCS&quot;  &quot;TCGA-THYM&quot; &quot;TCGA-COAD&quot; &quot;TCGA-ESCA&quot; &quot;TCGA-GBM&quot;  &quot;TCGA-KICH&quot; &quot;TCGA-HNSC&quot; &quot;TCGA-PRAD&quot;</span></span><br><span class="line"><span class="comment"># [21] &quot;TCGA-OV&quot;   &quot;TCGA-LUSC&quot; &quot;TCGA-LAML&quot; &quot;TCGA-LGG&quot;  &quot;TCGA-SARC&quot; &quot;TCGA-BRCA&quot; &quot;TCGA-READ&quot; &quot;TCGA-LUAD&quot; &quot;TCGA-STAD&quot; &quot;TCGA-THCA&quot;</span></span><br><span class="line"><span class="comment"># [31] &quot;TCGA-KIRC&quot; &quot;TCGA-SKCM&quot; &quot;TCGA-UCEC&quot;</span></span><br><span class="line">projects <span class="operator">%&gt;%</span> <span class="built_in">length</span></span><br><span class="line"><span class="comment"># [1] 33</span></span><br></pre></td></tr></table></figure>

<p>TCGA的33种癌症简写对应全称可在<a target="_blank" rel="noopener" href="https://xenabrowser.net/datapages/?host=https://gdc.xenahubs.net&removeHub=https://xena.treehouse.gi.ucsc.edu:443">UCSC Xena</a>查到.</p>
<p><img src="/img/20230323213327.png" alt="GDC Hub"></p>
<p>查看一个项目内的可下载文件:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line">TCGAbiolinks<span class="operator">:::</span>getProjectSummary<span class="punctuation">(</span><span class="string">&quot;TCGA-COAD&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># $file_count</span></span><br><span class="line"><span class="comment"># [1] 22585</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># $data_categories</span></span><br><span class="line"><span class="comment">#   file_count case_count               data_category</span></span><br><span class="line"><span class="comment"># 1       4007        460       Copy Number Variation</span></span><br><span class="line"><span class="comment"># 2       2932        460            Sequencing Reads</span></span><br><span class="line"><span class="comment"># 3       6482        434 Simple Nucleotide Variation</span></span><br><span class="line"><span class="comment"># 4       1665        457             DNA Methylation</span></span><br><span class="line"><span class="comment"># 5        995        461                    Clinical</span></span><br><span class="line"><span class="comment"># 6       1978        459     Transcriptome Profiling</span></span><br><span class="line"><span class="comment"># 7       2835        461                 Biospecimen</span></span><br><span class="line"><span class="comment"># 8        363        360          Proteome Profiling</span></span><br><span class="line"><span class="comment"># 9       1328        289        Structural Variation</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># $case_count</span></span><br><span class="line"><span class="comment"># [1] 461</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># $file_size</span></span><br><span class="line"><span class="comment"># [1] 7.004757e+13</span></span><br></pre></td></tr></table></figure>

<p>下载并提取mRNA的表达矩阵, 其中数据类别 <code>data.category</code> 是 <code>Transcriptome Profiling</code> 代表转录组数据; 数据类型 <code>data.type</code> 是 <code>Gene Expression Quantification</code> , 代表表达谱数据; 定量方式 <code>workflow.type</code> 是 <code>STAR - Counts</code> , 内含了 <code>Counts</code> &#x2F; <code>FPKM</code> 和 <code>FPKM-UQ</code> 三种数据类型.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 查询</span></span><br><span class="line">query <span class="operator">&lt;-</span> GDCquery<span class="punctuation">(</span></span><br><span class="line">    project <span class="operator">=</span> project<span class="punctuation">,</span></span><br><span class="line">    data.category <span class="operator">=</span> <span class="string">&quot;Transcriptome Profiling&quot;</span><span class="punctuation">,</span></span><br><span class="line">    data.type <span class="operator">=</span> <span class="string">&quot;Gene Expression Quantification&quot;</span><span class="punctuation">,</span></span><br><span class="line">    workflow.type <span class="operator">=</span> <span class="string">&quot;STAR - Counts&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">head<span class="punctuation">(</span>query<span class="operator">$</span>results<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">,</span> n<span class="operator">=</span><span class="number">10</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> View<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 下载</span></span><br><span class="line">GDCdownload<span class="punctuation">(</span>query<span class="punctuation">,</span> method <span class="operator">=</span> <span class="string">&quot;api&quot;</span><span class="punctuation">,</span> files.per.chunk <span class="operator">=</span> <span class="number">100</span><span class="punctuation">)</span> <span class="comment"># 每次下载100个文件</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 整理</span></span><br><span class="line"><span class="comment"># GDCprepare(query, save = T, save.filename = paste0(project, &quot;_mRNA.Rdata&quot;))</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p><img src="/img/20230323213227.png"></p>
<p>这里出现了一个意外… 我的硬盘提示空间不足了.. 在这个内存动不动64的年代, 我这个硬盘总共200g的可怜人实惨.. 可见 <code>GDCprepare</code> 函数需要强大的内存和硬盘空间, 我的本地电脑是做不到的, 因此继续使用老方案进行数据处理. 目前为止, 通过 <code>TCGAbiolinks</code> 进行数据下载的目的已经圆满达到.</p>
<p><img src="/img/20230323220222.png"></p>
<p>首先保存 <code>MANIFEST</code> 文件以供后续分析:</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br></pre></td><td class="code"><pre><span class="line">shelfEnvironment<span class="punctuation">(</span>paste<span class="punctuation">(</span>imput_dir<span class="punctuation">,</span> <span class="string">&quot;GDCdata&quot;</span><span class="punctuation">,</span> project<span class="punctuation">,</span> sep <span class="operator">=</span> <span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> path <span class="operator">=</span> root_dir<span class="punctuation">)</span></span><br><span class="line">write.csv<span class="punctuation">(</span>query<span class="operator">$</span>results<span class="punctuation">[[</span><span class="number">1</span><span class="punctuation">]</span><span class="punctuation">]</span><span class="punctuation">,</span> file <span class="operator">=</span> paste0<span class="punctuation">(</span>project<span class="punctuation">,</span> <span class="string">&quot;_MANIFEST.csv&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p><code>shelfEnvironment</code> 函数来源于 <code>obgetDEGs</code> 包, 可使用 <code>devtools::install_github(&#39;sandy9707/obgetDEGs&#39;)</code> 命令安装, 函数的作用是将目标文件夹设定为工作目录, 如果该目录不存在便创建. 详情可参考<a target="_blank" rel="noopener" href="https://github.com/sandy9707/obgetDEGs">GitHub - sandy9707&#x2F;obgetDEGs</a>.</p>
<p> 该函数的应用场景是：当需要在R中读取或写入数据时，需要指定存储数据的文件夹路径。但在执行R代码时，可能需要将当前工作目录更改为存储数据的文件夹路径。如果文件夹不存在，需要创建文件夹。这时， shelfEnvironment 函数可以帮助我们检查并创建文件夹，使得数据可以正常读取或写入。</p>
<h3 id="表达谱数据处理"><a href="#表达谱数据处理" class="headerlink" title="表达谱数据处理"></a>表达谱数据处理</h3><p>清空环境, 读取MANIFEST信息, 特别是需要样本名和文件夹名.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># !整理----</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 清除当前环境中的所有对象</span></span><br><span class="line"></span><br><span class="line">rm<span class="punctuation">(</span><span class="built_in">list</span> <span class="operator">=</span> ls<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment">## 设置主文件夹路径, 并设置工作目录</span></span><br><span class="line"></span><br><span class="line"><span class="punctuation">(</span>root_dir <span class="operator">&lt;-</span> sub<span class="punctuation">(</span><span class="string">&quot;/code.+&quot;</span><span class="punctuation">,</span> <span class="string">&quot;&quot;</span><span class="punctuation">,</span> rstudioapi<span class="operator">::</span>getSourceEditorContext<span class="punctuation">(</span><span class="punctuation">)</span><span class="operator">$</span>path<span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">source<span class="punctuation">(</span>paste<span class="punctuation">(</span>root_dir<span class="punctuation">,</span> <span class="string">&quot;code&quot;</span><span class="punctuation">,</span> <span class="string">&quot;prepare.R&quot;</span><span class="punctuation">,</span> sep <span class="operator">=</span> <span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">project <span class="operator">&lt;-</span> <span class="string">&quot;TCGA-COAD&quot;</span></span><br><span class="line">shelfEnvironment<span class="punctuation">(</span>paste<span class="punctuation">(</span>imput_dir<span class="punctuation">,</span> <span class="string">&quot;GDCdata&quot;</span><span class="punctuation">,</span> project<span class="punctuation">,</span> sep <span class="operator">=</span> <span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> path <span class="operator">=</span> root_dir<span class="punctuation">)</span></span><br><span class="line">json <span class="operator">&lt;-</span> read.csv<span class="punctuation">(</span>paste0<span class="punctuation">(</span>project<span class="punctuation">,</span> <span class="string">&quot;_MANIFEST.csv&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># json[, c(&quot;cases&quot;, &quot;file_id&quot;)]</span></span><br><span class="line">case_names <span class="operator">&lt;-</span> json<span class="punctuation">[</span><span class="punctuation">,</span> <span class="string">&quot;cases&quot;</span><span class="punctuation">]</span></span><br><span class="line">filedir_in_json <span class="operator">&lt;-</span> json<span class="punctuation">[</span><span class="punctuation">,</span> <span class="string">&quot;file_id&quot;</span><span class="punctuation">]</span></span><br></pre></td></tr></table></figure>

<p>选择提取部分</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 提取表达量至一个数据框(以tibble格式),counts值选4,fpkm选8,tpm选7</span></span><br><span class="line">extract_type <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;counts&quot;</span><span class="punctuation">,</span> <span class="string">&quot;fpkm&quot;</span><span class="punctuation">,</span> <span class="string">&quot;tpm&quot;</span><span class="punctuation">)</span><span class="punctuation">[</span><span class="number">1</span><span class="punctuation">]</span></span><br><span class="line">extract_num <span class="operator">&lt;-</span> <span class="built_in">switch</span><span class="punctuation">(</span>extract_type<span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;counts&quot;</span> <span class="operator">=</span> <span class="number">4</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;fpkm&quot;</span> <span class="operator">=</span> <span class="number">8</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;tpm&quot;</span> <span class="operator">=</span> <span class="number">7</span></span><br><span class="line"><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>开始提取, 原理是进入每一个文件夹并提取某列, 再结合基因类型, 并去重.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br><span class="line">22</span><br><span class="line">23</span><br><span class="line">24</span><br><span class="line">25</span><br><span class="line">26</span><br><span class="line">27</span><br><span class="line">28</span><br><span class="line">29</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 开始提取</span></span><br><span class="line">matrix_MMRF_list <span class="operator">&lt;-</span> <span class="built_in">list</span><span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">shelfEnvironment<span class="punctuation">(</span>paste<span class="punctuation">(</span>imput_dir<span class="punctuation">,</span> <span class="string">&quot;GDCdata&quot;</span><span class="punctuation">,</span> project<span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;/harmonized/Transcriptome_Profiling/Gene_Expression_Quantification&quot;</span><span class="punctuation">,</span></span><br><span class="line">    sep <span class="operator">=</span> <span class="string">&quot;/&quot;</span></span><br><span class="line"><span class="punctuation">)</span><span class="punctuation">,</span> path <span class="operator">=</span> root_dir<span class="punctuation">)</span></span><br><span class="line"><span class="keyword">for</span> <span class="punctuation">(</span>i <span class="keyword">in</span> <span class="number">1</span><span class="operator">:</span><span class="built_in">length</span><span class="punctuation">(</span>filedir_in_json<span class="punctuation">)</span><span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    setwd<span class="punctuation">(</span>paste0<span class="punctuation">(</span><span class="string">&quot;./&quot;</span><span class="punctuation">,</span> filedir_in_json<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">    matrix_MMRF_list<span class="punctuation">[[</span>i<span class="punctuation">]</span><span class="punctuation">]</span> <span class="operator">&lt;-</span> read.table<span class="punctuation">(</span></span><br><span class="line">        file <span class="operator">=</span> list.files<span class="punctuation">(</span>pattern <span class="operator">=</span> <span class="string">&quot;.tsv&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">        header <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span> fill <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line">    <span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span> extract_num<span class="punctuation">]</span></span><br><span class="line">    setwd<span class="punctuation">(</span><span class="string">&quot;../&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span></span><br><span class="line">setwd<span class="punctuation">(</span>paste0<span class="punctuation">(</span><span class="string">&quot;./&quot;</span><span class="punctuation">,</span> filedir_in_json<span class="punctuation">[</span>i<span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">probematrix <span class="operator">&lt;-</span> read.table<span class="punctuation">(</span>list.files<span class="punctuation">(</span>pattern <span class="operator">=</span> <span class="string">&quot;.tsv&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    header <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">,</span> fill <span class="operator">=</span> <span class="literal">TRUE</span></span><br><span class="line"><span class="punctuation">)</span><span class="punctuation">[</span><span class="punctuation">,</span> <span class="number">2</span><span class="operator">:</span><span class="number">3</span><span class="punctuation">]</span></span><br><span class="line">setwd<span class="punctuation">(</span><span class="string">&quot;../&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line">matrix_MMRF <span class="operator">&lt;-</span> do.call<span class="punctuation">(</span>cbind<span class="punctuation">,</span> matrix_MMRF_list<span class="punctuation">)</span></span><br><span class="line">matrix_MMRF <span class="operator">&lt;-</span> cbind<span class="punctuation">(</span>probematrix<span class="punctuation">,</span> matrix_MMRF<span class="punctuation">)</span></span><br><span class="line">tibble_MMRF <span class="operator">&lt;-</span> tibble<span class="operator">::</span>as_tibble<span class="punctuation">(</span>matrix_MMRF<span class="punctuation">)</span></span><br><span class="line">colnames<span class="punctuation">(</span>tibble_MMRF<span class="punctuation">)</span> <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;gene_name&quot;</span><span class="punctuation">,</span> <span class="string">&quot;gene_type&quot;</span><span class="punctuation">,</span> str_sub<span class="punctuation">(</span>case_names<span class="punctuation">,</span> <span class="number">1</span><span class="punctuation">,</span> <span class="number">16</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="comment">#</span></span><br><span class="line">duplicated<span class="punctuation">(</span>colnames<span class="punctuation">(</span>tibble_MMRF<span class="punctuation">)</span><span class="punctuation">,</span> fromLast <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> table<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">tibble_MMRF <span class="operator">&lt;-</span> tibble_MMRF<span class="punctuation">[</span><span class="punctuation">,</span> <span class="operator">!</span>duplicated<span class="punctuation">(</span>colnames<span class="punctuation">(</span>tibble_MMRF<span class="punctuation">)</span><span class="punctuation">,</span> fromLast <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span><span class="punctuation">]</span></span><br><span class="line">duplicated<span class="punctuation">(</span>colnames<span class="punctuation">(</span>tibble_MMRF<span class="punctuation">)</span><span class="punctuation">,</span> fromLast <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span> table<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>提取蛋白编码基因并将基因名保留转换行名.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br></pre></td><td class="code"><pre><span class="line"></span><br><span class="line"><span class="comment"># 提取蛋白编码基因</span></span><br><span class="line">pcg <span class="operator">&lt;-</span> <span class="built_in">c</span><span class="punctuation">(</span></span><br><span class="line">    <span class="string">&quot;protein_coding&quot;</span><span class="punctuation">,</span> <span class="string">&quot;IG_V_gene&quot;</span><span class="punctuation">,</span> <span class="string">&quot;IG_D_gene&quot;</span><span class="punctuation">,</span> <span class="string">&quot;IG_J_gene&quot;</span><span class="punctuation">,</span></span><br><span class="line">    <span class="string">&quot;IG_C_gene&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TR_V_gene&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TR_D_gene&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TR_J_gene&quot;</span><span class="punctuation">,</span> <span class="string">&quot;TR_C_gene&quot;</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"><span class="comment"># 创建一个tibble_MMRF对象，使用dplyr::filter()方法筛选出gene_type包含于pcg的所有行</span></span><br><span class="line">mrna_exprset <span class="operator">&lt;-</span> tibble_MMRF <span class="operator">%&gt;%</span></span><br><span class="line">    dplyr<span class="operator">::</span>filter<span class="punctuation">(</span>gene_type <span class="operator">%in%</span> pcg<span class="punctuation">)</span> <span class="operator">%&gt;%</span></span><br><span class="line">    <span class="comment"># 使用dplyr::select()方法去掉gene_type列</span></span><br><span class="line">    dplyr<span class="operator">::</span>select<span class="punctuation">(</span><span class="operator">-</span>gene_type<span class="punctuation">)</span> <span class="operator">%&gt;%</span></span><br><span class="line">    <span class="comment"># 使用dplyr::distinct()方法去除重复的行，保留第一次出现的行</span></span><br><span class="line">    dplyr<span class="operator">::</span>distinct<span class="punctuation">(</span>gene_name<span class="punctuation">,</span> .keep_all <span class="operator">=</span> <span class="literal">TRUE</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span></span><br><span class="line">    <span class="comment"># 使用tibble::column_to_rownames()方法将gene_name列转换为行名</span></span><br><span class="line">    tibble<span class="operator">::</span>column_to_rownames<span class="punctuation">(</span><span class="string">&quot;gene_name&quot;</span><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<p>通过TCGA样本命名规则筛选需求样本并将对照组前置.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 查看去掉01A和11A的样本个数, 通过数量可以看出效果一致</span></span><br><span class="line">mrna_exprset <span class="operator">%&gt;%</span></span><br><span class="line">    select<span class="punctuation">(</span><span class="operator">-</span>matches<span class="punctuation">(</span><span class="string">&quot;[^(01)]A$|[^(11)]A$&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span></span><br><span class="line">    ncol<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line">mrna_exprset <span class="operator">%&gt;%</span></span><br><span class="line">    select<span class="punctuation">(</span>matches<span class="punctuation">(</span><span class="string">&quot;[01]1[A]$&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span></span><br><span class="line">    ncol<span class="punctuation">(</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 筛选, 只要01A和11A的样本</span></span><br><span class="line"><span class="comment"># 重新排序,将癌旁排在前面便于下一步筛选,0-9为癌数据,排在后面</span></span><br><span class="line">mrna_exprset <span class="operator">&lt;-</span> mrna_exprset <span class="operator">%&gt;%</span></span><br><span class="line">    select<span class="punctuation">(</span>matches<span class="punctuation">(</span><span class="string">&quot;[01]1A$&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span> <span class="operator">%&gt;%</span></span><br><span class="line">    select<span class="punctuation">(</span><span class="operator">-</span>matches<span class="punctuation">(</span><span class="string">&quot;.-0[1-9][A]$&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> everything<span class="punctuation">(</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line">ncol<span class="punctuation">(</span>mrna_exprset<span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<p>写出表达矩阵, 特征列表和分组列表.</p>
<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br><span class="line">19</span><br><span class="line">20</span><br><span class="line">21</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 写出表达矩阵_extract_type</span></span><br><span class="line">shelfEnvironment<span class="punctuation">(</span>paste<span class="punctuation">(</span>imput_dir<span class="punctuation">,</span> <span class="string">&quot;GDCdata&quot;</span><span class="punctuation">,</span> project<span class="punctuation">,</span> sep <span class="operator">=</span> <span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    path <span class="operator">=</span> root_dir</span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">write.csv<span class="punctuation">(</span>mrna_exprset<span class="punctuation">,</span> paste0<span class="punctuation">(</span><span class="string">&quot;Eset_&quot;</span><span class="punctuation">,</span> extract_type<span class="punctuation">,</span> <span class="string">&quot;_ProCoding.csv&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> row.names <span class="operator">=</span> <span class="built_in">T</span><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 写出特征列表, 也就是基因名</span></span><br><span class="line">write.table<span class="punctuation">(</span>rownames<span class="punctuation">(</span>mrna_exprset<span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;feature_list.csv&quot;</span><span class="punctuation">,</span></span><br><span class="line">    row.names <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> <span class="built_in">quote</span> <span class="operator">=</span> <span class="literal">FALSE</span><span class="punctuation">,</span> col.names <span class="operator">=</span> <span class="literal">FALSE</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br><span class="line"><span class="comment"># 写出分组列表, 也就是组别</span></span><br><span class="line">samples <span class="operator">&lt;-</span> colnames<span class="punctuation">(</span>mrna_exprset<span class="punctuation">)</span></span><br><span class="line">group_definition_matrix <span class="operator">&lt;-</span> data.frame<span class="punctuation">(</span></span><br><span class="line">    samples<span class="punctuation">,</span></span><br><span class="line">    group <span class="operator">=</span> ifelse<span class="punctuation">(</span>grepl<span class="punctuation">(</span><span class="string">&quot;11A$&quot;</span><span class="punctuation">,</span> samples<span class="punctuation">)</span><span class="punctuation">,</span> <span class="string">&quot;control&quot;</span><span class="punctuation">,</span> <span class="string">&quot;test&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">write.csv<span class="punctuation">(</span>group_definition_matrix<span class="punctuation">,</span></span><br><span class="line">    file <span class="operator">=</span> <span class="string">&quot;group_definition_matrix.csv&quot;</span><span class="punctuation">,</span> row.names <span class="operator">=</span> <span class="built_in">F</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line"></span><br></pre></td></tr></table></figure>

<figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br></pre></td><td class="code"><pre><span class="line">sapply<span class="punctuation">(</span>projects<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>project<span class="punctuation">)</span><span class="punctuation">&#123;</span></span><br><span class="line">  </span><br><span class="line">  query <span class="operator">&lt;-</span> GDCquery<span class="punctuation">(</span>project <span class="operator">=</span> project<span class="punctuation">,</span></span><br><span class="line">                    data.category <span class="operator">=</span> <span class="string">&quot;Clinical&quot;</span><span class="punctuation">,</span> </span><br><span class="line">                    file.type <span class="operator">=</span> <span class="string">&quot;xml&quot;</span><span class="punctuation">)</span></span><br><span class="line">  GDCdownload<span class="punctuation">(</span>query<span class="punctuation">)</span></span><br><span class="line">  clinical <span class="operator">&lt;-</span> GDCprepare_clinic<span class="punctuation">(</span>query<span class="punctuation">,</span> clinical.info <span class="operator">=</span> <span class="string">&quot;patient&quot;</span><span class="punctuation">)</span></span><br><span class="line">  saveRDS<span class="punctuation">(</span>clinical<span class="punctuation">,</span>file <span class="operator">=</span> paste0<span class="punctuation">(</span>project<span class="punctuation">,</span><span class="string">&quot;_clinical.rds&quot;</span><span class="punctuation">)</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="临床数据下载及处理"><a href="#临床数据下载及处理" class="headerlink" title="临床数据下载及处理"></a>临床数据下载及处理</h2><figure class="highlight r"><table><tr><td class="gutter"><pre><span class="line">1</span><br><span class="line">2</span><br><span class="line">3</span><br><span class="line">4</span><br><span class="line">5</span><br><span class="line">6</span><br><span class="line">7</span><br><span class="line">8</span><br><span class="line">9</span><br><span class="line">10</span><br><span class="line">11</span><br><span class="line">12</span><br><span class="line">13</span><br><span class="line">14</span><br><span class="line">15</span><br><span class="line">16</span><br><span class="line">17</span><br><span class="line">18</span><br></pre></td><td class="code"><pre><span class="line"><span class="comment"># 临床数据</span></span><br><span class="line">project</span><br><span class="line">sapply<span class="punctuation">(</span>projects<span class="punctuation">,</span> <span class="keyword">function</span><span class="punctuation">(</span>project<span class="punctuation">)</span> <span class="punctuation">&#123;</span></span><br><span class="line">    query <span class="operator">&lt;-</span> GDCquery<span class="punctuation">(</span></span><br><span class="line">        project <span class="operator">=</span> project<span class="punctuation">,</span></span><br><span class="line">        data.category <span class="operator">=</span> <span class="string">&quot;Clinical&quot;</span><span class="punctuation">,</span></span><br><span class="line">        file.type <span class="operator">=</span> <span class="string">&quot;xml&quot;</span></span><br><span class="line">    <span class="punctuation">)</span></span><br><span class="line">    GDCdownload<span class="punctuation">(</span>query<span class="punctuation">)</span></span><br><span class="line">    clinical <span class="operator">&lt;-</span> GDCprepare_clinic<span class="punctuation">(</span>query<span class="punctuation">,</span> clinical.info <span class="operator">=</span> <span class="string">&quot;patient&quot;</span><span class="punctuation">)</span></span><br><span class="line">    <span class="comment"># saveRDS(clinical, file = paste0(project, &quot;_clinical.rds&quot;))</span></span><br><span class="line"><span class="punctuation">&#125;</span><span class="punctuation">)</span></span><br><span class="line">write.csv<span class="punctuation">(</span>clinical<span class="punctuation">,</span> paste0<span class="punctuation">(</span><span class="string">&quot;Phenodata.csv&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> row.names <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br><span class="line">col_survivaldata <span class="operator">&lt;-</span> setNames<span class="punctuation">(</span></span><br><span class="line">    data.frame<span class="punctuation">(</span>clinical<span class="punctuation">[</span><span class="punctuation">,</span> <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;bcr_patient_barcode&quot;</span><span class="punctuation">,</span> <span class="string">&quot;vital_status&quot;</span><span class="punctuation">,</span> <span class="string">&quot;days_to_death&quot;</span><span class="punctuation">)</span><span class="punctuation">]</span><span class="punctuation">)</span><span class="punctuation">,</span></span><br><span class="line">    <span class="built_in">c</span><span class="punctuation">(</span><span class="string">&quot;patient&quot;</span><span class="punctuation">,</span> <span class="string">&quot;status&quot;</span><span class="punctuation">,</span> <span class="string">&quot;time&quot;</span><span class="punctuation">)</span></span><br><span class="line"><span class="punctuation">)</span></span><br><span class="line">write.csv<span class="punctuation">(</span>col_survivaldata<span class="punctuation">,</span> paste<span class="punctuation">(</span><span class="string">&quot;Survivalata.csv&quot;</span><span class="punctuation">,</span> sep <span class="operator">=</span> <span class="string">&quot;/&quot;</span><span class="punctuation">)</span><span class="punctuation">,</span> row.names <span class="operator">=</span> <span class="built_in">F</span><span class="punctuation">)</span></span><br></pre></td></tr></table></figure>

<h2 id="总结"><a href="#总结" class="headerlink" title="总结"></a>总结</h2><p>如想获得本文的全套代码以及工作环境, 只需要在公众号私信发送 <code>TCGA数据下载整理</code> 就可以获得.</p>
<h2 id="引用"><a href="#引用" class="headerlink" title="引用"></a>引用</h2><ol>
<li><a target="_blank" rel="noopener" href="https://portal.gdc.cancer.gov/">GDC官网</a></li>
<li><a target="_blank" rel="noopener" href="https://docs.gdc.cancer.gov/Data_Transfer_Tool/Users_Guide/Getting_Started/">官方下载工具gdc-client</a></li>
<li><a target="_blank" rel="noopener" href="https://xena.ucsc.edu/">UCSC Xena</a></li>
<li><a target="_blank" rel="noopener" href="https://gdac.broadinstitute.org/">Firehouse</a></li>
<li><a target="_blank" rel="noopener" href="https://bioconductor.org/packages/release/bioc/html/TCGAbiolinks.html">TCGAbiolinks</a></li>
<li><a target="_blank" rel="noopener" href="https://cloud.tencent.com/developer/article/1424598">TCGA数据下载—TCGAbiolinks包参数详解 - 腾讯云开发者社区-腾讯云</a></li>
<li><a target="_blank" rel="noopener" href="https://mdnice.com/writing/43c4248612b14223840b469a1fdf7a2f">新版TCGAbiolinks包学习：批量下载数据新版TCGA数据 - mdnice 墨滴</a></li>
<li><a target="_blank" rel="noopener" href="https://blog.csdn.net/weixin_41368414/article/details/122922782">TCGA &#x2F; 癌症简称 &#x2F; 缩写 &#x2F; TCGA癌症中英文对照</a></li>
<li><a target="_blank" rel="noopener" href="https://github.com/sandy9707/obgetDEGs">GitHub - sandy9707&#x2F;obgetDEGs</a></li>
<li><a target="_blank" rel="noopener" href="https://www.jianshu.com/p/bf5eaf8230d8">TCGA样本命名规则</a></li>
</ol>
</article><div class="post-copyright"><div class="post-copyright__author"><span class="post-copyright-meta">Author: </span><span class="post-copyright-info"><a href="https://yeyeziblog.eu.org">Yezi</a></span></div><div class="post-copyright__type"><span class="post-copyright-meta">Link: </span><span class="post-copyright-info"><a href="https://yeyeziblog.eu.org/2023/03/23/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/TCGA/TCGAbiolinks%E5%8C%85%E5%8F%82%E6%95%B0%E8%AF%A6%E8%A7%A3/">https://yeyeziblog.eu.org/2023/03/23/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/TCGA/TCGAbiolinks%E5%8C%85%E5%8F%82%E6%95%B0%E8%AF%A6%E8%A7%A3/</a></span></div><div class="post-copyright__notice"><span class="post-copyright-meta">Copyright Notice: </span><span class="post-copyright-info">All articles in this blog are licensed under <a target="_blank" rel="noopener" href="https://creativecommons.org/licenses/by-nc-sa/4.0/">CC BY-NC-SA 4.0</a> unless stating additionally.</span></div></div><div class="tag_share"><div class="post-meta__tag-list"><a class="post-meta__tags" href="/tags/R-%E8%AF%AD%E8%A8%80/">R 语言</a><a class="post-meta__tags" href="/tags/TCGA/">TCGA</a></div><div class="post_share"><div class="social-share" data-image="/img/TCGA%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD_2023%E5%B9%B403%E6%9C%88.jpg" data-sites="facebook,twitter,wechat,weibo,qq"></div><link rel="stylesheet" href="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/css/share.min.css" media="print" onload="this.media='all'"><script src="https://cdn.jsdelivr.net/npm/butterfly-extsrc/sharejs/dist/js/social-share.min.js" defer></script></div></div><nav class="pagination-post" id="pagination"><div class="prev-post pull-left"><a href="/2023/03/27/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90%E5%88%86%E7%BB%84%E6%9E%84%E5%BB%BA%E5%88%B0%E5%BA%95%E8%B0%81%E5%9C%A8%E5%89%8D%E9%9D%A2--%E5%85%B3%E4%BA%8Elimma%E5%8C%85%E4%B8%ADmodel.matrix()%E7%9A%84%E9%97%AE%E9%A2%98/" title="差异分析分组构建到底谁在前面--关于limma包中model.matrix()的问题"><div class="cover" style="background: var(--default-bg-color)"></div><div class="pagination-info"><div class="label">Previous Post</div><div class="prev_info">差异分析分组构建到底谁在前面--关于limma包中model.matrix()的问题</div></div></a></div><div class="next-post pull-right"><a href="/2023/03/21/Tech/markdown/%E5%A6%82%E4%BD%95%E7%94%A8Markdown%E5%86%99%E6%AF%95%E4%B8%9A%E8%AE%BA%E6%96%87/" title="如何用 Markdown 写毕业论文"><div class="cover" style="background: var(--default-bg-color)"></div><div class="pagination-info"><div class="label">Next Post</div><div class="next_info">如何用 Markdown 写毕业论文</div></div></a></div></nav><div class="relatedPosts"><div class="headline"><i class="fas fa-thumbs-up fa-fw"></i><span>Related Articles</span></div><div class="relatedPosts-list"><div><a href="/2023/05/09/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/TCGA/TCGA-miRNA%E6%95%B0%E6%8D%AE%E6%95%B4%E7%90%86/" title="TCGA-miRNA数据整理"><img class="cover" src="/img/TCGA-miRNA%E6%95%B0%E6%8D%AE%E6%95%B4%E7%90%86_2023-05-14-16-06-18.png" alt="cover"><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2023-05-09</div><div class="title">TCGA-miRNA数据整理</div></div></a></div><div><a href="/2023/09/25/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/%E4%B8%80%E7%A7%8D%E8%BE%83%E4%B8%BA%E9%80%9A%E7%94%A8%E7%9A%84GEO%E6%95%B0%E6%8D%AEID%E8%BD%AC%E6%8D%A2%E7%9A%84%E6%96%B9%E6%A1%88/" title="一种较为通用的GEO数据ID转换的方案"><div class="cover" style="background: var(--default-bg-color)"></div><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2023-09-25</div><div class="title">一种较为通用的GEO数据ID转换的方案</div></div></a></div><div><a href="/2023/03/27/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/%E5%B7%AE%E5%BC%82%E5%88%86%E6%9E%90%E5%88%86%E7%BB%84%E6%9E%84%E5%BB%BA%E5%88%B0%E5%BA%95%E8%B0%81%E5%9C%A8%E5%89%8D%E9%9D%A2--%E5%85%B3%E4%BA%8Elimma%E5%8C%85%E4%B8%ADmodel.matrix()%E7%9A%84%E9%97%AE%E9%A2%98/" title="差异分析分组构建到底谁在前面--关于limma包中model.matrix()的问题"><div class="cover" style="background: var(--default-bg-color)"></div><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2023-03-27</div><div class="title">差异分析分组构建到底谁在前面--关于limma包中model.matrix()的问题</div></div></a></div><div><a href="/2023/03/08/Tech/VS%20CODE/%E6%9B%B4%E7%8E%B0%E4%BB%A3%E7%9A%84R%E4%BB%A3%E7%A0%81%E7%BC%96%E8%BE%91%E5%99%A8----%E6%9C%AC%E5%9C%B0VSCode-R%E6%90%AD%E5%BB%BA/" title="更现代的 R 代码编辑器----本地 VSCode-R 搭建"><div class="cover" style="background: var(--default-bg-color)"></div><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2023-03-08</div><div class="title">更现代的 R 代码编辑器----本地 VSCode-R 搭建</div></div></a></div><div><a href="/2023/05/24/Tech/Linux/%E4%BB%A3%E7%90%86%E5%8F%AF%E7%94%A8%E6%80%A7%E6%B5%8B%E8%AF%95%E5%8F%8A%E4%BD%BF%E7%94%A8/" title="代理可用性测试及使用"><div class="cover" style="background: var(--default-bg-color)"></div><div class="content is-center"><div class="date"><i class="far fa-calendar-alt fa-fw"></i> 2023-05-24</div><div class="title">代理可用性测试及使用</div></div></a></div></div></div><hr/><div id="post-comment"><div class="comment-head"><div class="comment-headline"><i class="fas fa-comments fa-fw"></i><span> Comment</span></div></div><div class="comment-wrap"><div><div id="gitalk-container"></div></div></div></div></div><div class="aside-content" id="aside-content"><div class="card-widget card-info"><div class="is-center"><div class="avatar-img"><img src="/image/avatar.png" onerror="this.onerror=null;this.src='/img/friend_404.gif'" alt="avatar"/></div><div class="author-info__name">Yezi</div><div class="author-info__description"></div></div><div class="card-info-data site-data is-center"><a href="/archives/"><div class="headline">Articles</div><div class="length-num">152</div></a><a href="/tags/"><div class="headline">Tags</div><div class="length-num">91</div></a><a href="/categories/"><div class="headline">Categories</div><div class="length-num">42</div></a></div><a id="card-info-btn" target="_blank" rel="noopener" href="https://github.com/sandy9707"><i class="fab fa-github"></i><span>Follow Me</span></a></div><div class="card-widget card-announcement"><div class="item-headline"><i class="fas fa-bullhorn fa-shake"></i><span>Announcement</span></div><div class="announcement_content">This is my Blog</div></div><div class="sticky_layout"><div class="card-widget" id="card-toc"><div class="item-headline"><i class="fas fa-stream"></i><span>Catalog</span><span class="toc-percentage"></span></div><div class="toc-content"><ol class="toc"><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%BC%95%E8%A8%80"><span class="toc-number">1.</span> <span class="toc-text">引言</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%95%88%E6%9E%9C%E5%B1%95%E7%A4%BA"><span class="toc-number">2.</span> <span class="toc-text">效果展示</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E8%BF%87%E7%A8%8B"><span class="toc-number">3.</span> <span class="toc-text">过程</span></a><ol class="toc-child"><li class="toc-item toc-level-3"><a class="toc-link" href="#%E4%B8%8B%E8%BD%BD"><span class="toc-number">3.1.</span> <span class="toc-text">下载</span></a></li><li class="toc-item toc-level-3"><a class="toc-link" href="#%E8%A1%A8%E8%BE%BE%E8%B0%B1%E6%95%B0%E6%8D%AE%E5%A4%84%E7%90%86"><span class="toc-number">3.2.</span> <span class="toc-text">表达谱数据处理</span></a></li></ol></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E4%B8%B4%E5%BA%8A%E6%95%B0%E6%8D%AE%E4%B8%8B%E8%BD%BD%E5%8F%8A%E5%A4%84%E7%90%86"><span class="toc-number">4.</span> <span class="toc-text">临床数据下载及处理</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E6%80%BB%E7%BB%93"><span class="toc-number">5.</span> <span class="toc-text">总结</span></a></li><li class="toc-item toc-level-2"><a class="toc-link" href="#%E5%BC%95%E7%94%A8"><span class="toc-number">6.</span> <span class="toc-text">引用</span></a></li></ol></div></div><div class="card-widget card-recent-post"><div class="item-headline"><i class="fas fa-history"></i><span>Recent Post</span></div><div class="aside-list"><div class="aside-list-item"><a class="thumbnail" href="/2024/11/02/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/R/rserver%E4%B8%AD%E6%97%A0%E6%B3%95%E4%BD%BF%E7%94%A8conda/" title="rserver 中无法使用 conda 的解决办法"><img src="/img/%E5%8F%B6%E5%AD%90%E7%9A%84%E7%A5%9E%E5%A5%87%E5%B0%8F%E8%BD%AF%E4%BB%B6.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="rserver 中无法使用 conda 的解决办法"/></a><div class="content"><a class="title" href="/2024/11/02/%E7%94%9F%E7%89%A9%E4%BF%A1%E6%81%AF%E5%AD%A6/R/rserver%E4%B8%AD%E6%97%A0%E6%B3%95%E4%BD%BF%E7%94%A8conda/" title="rserver 中无法使用 conda 的解决办法">rserver 中无法使用 conda 的解决办法</a><time datetime="2024-11-02T12:28:59.000Z" title="Created 2024-11-02 20:28:59">2024-11-02</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/11/02/%E6%9D%82%E8%B0%88/%E5%90%8C%E4%BA%BA%E6%96%87/%E8%B5%B0%E8%BF%9B%E4%BF%AE%E4%BB%992024%E5%90%8C%E4%BA%BA%E6%B4%BB%E5%8A%A8/" title="走进修仙 2024 同人活动"><img src="/img/%E8%B5%B0%E8%BF%9B%E4%BF%AE%E4%BB%99.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="走进修仙 2024 同人活动"/></a><div class="content"><a class="title" href="/2024/11/02/%E6%9D%82%E8%B0%88/%E5%90%8C%E4%BA%BA%E6%96%87/%E8%B5%B0%E8%BF%9B%E4%BF%AE%E4%BB%992024%E5%90%8C%E4%BA%BA%E6%B4%BB%E5%8A%A8/" title="走进修仙 2024 同人活动">走进修仙 2024 同人活动</a><time datetime="2024-11-02T04:12:00.000Z" title="Created 2024-11-02 12:12:00">2024-11-02</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/11/01/Tech/Linux/%E5%AE%89%E8%A3%85nvidia-container-toolkit%E4%BB%A5%E6%94%AF%E6%8C%81docker%E4%BD%BF%E7%94%A8%E6%98%BE%E5%8D%A1/" title="安装 nvidia-container-toolkit 以支持 docker 使用显卡"><img src="/img/%E5%8F%B6%E5%AD%90%E7%9A%84%E7%A5%9E%E5%A5%87%E5%B0%8F%E8%BD%AF%E4%BB%B6.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="安装 nvidia-container-toolkit 以支持 docker 使用显卡"/></a><div class="content"><a class="title" href="/2024/11/01/Tech/Linux/%E5%AE%89%E8%A3%85nvidia-container-toolkit%E4%BB%A5%E6%94%AF%E6%8C%81docker%E4%BD%BF%E7%94%A8%E6%98%BE%E5%8D%A1/" title="安装 nvidia-container-toolkit 以支持 docker 使用显卡">安装 nvidia-container-toolkit 以支持 docker 使用显卡</a><time datetime="2024-11-01T01:47:45.000Z" title="Created 2024-11-01 09:47:45">2024-11-01</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/10/29/Tech/%E5%8F%B6%E5%AD%90%E7%9A%84%E6%8E%A8%E8%8D%90%E5%B0%8F%E8%BD%AF%E4%BB%B6/%E7%BA%AF%E5%9B%BEppt%E7%9A%84%E4%BC%98%E5%8A%BF%E5%92%8C%E8%BD%AC%E6%8D%A2%E6%96%B9%E6%B3%95/" title="纯图 ppt 的优势和转换方法"><img src="/img/%E5%8F%B6%E5%AD%90%E7%9A%84%E7%A5%9E%E5%A5%87%E5%B0%8F%E8%BD%AF%E4%BB%B6.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="纯图 ppt 的优势和转换方法"/></a><div class="content"><a class="title" href="/2024/10/29/Tech/%E5%8F%B6%E5%AD%90%E7%9A%84%E6%8E%A8%E8%8D%90%E5%B0%8F%E8%BD%AF%E4%BB%B6/%E7%BA%AF%E5%9B%BEppt%E7%9A%84%E4%BC%98%E5%8A%BF%E5%92%8C%E8%BD%AC%E6%8D%A2%E6%96%B9%E6%B3%95/" title="纯图 ppt 的优势和转换方法">纯图 ppt 的优势和转换方法</a><time datetime="2024-10-29T14:09:48.000Z" title="Created 2024-10-29 22:09:48">2024-10-29</time></div></div><div class="aside-list-item"><a class="thumbnail" href="/2024/10/25/Tech/Linux/centos%E4%BD%BF%E7%94%A8telnet%E8%BF%9E%E6%8E%A5/" title="centos 使用 telnet 连接"><img src="/img/%E5%8F%B6%E5%AD%90%E7%9A%84%E7%A5%9E%E5%A5%87%E5%B0%8F%E8%BD%AF%E4%BB%B6.jpg" onerror="this.onerror=null;this.src='/img/404.jpg'" alt="centos 使用 telnet 连接"/></a><div class="content"><a class="title" href="/2024/10/25/Tech/Linux/centos%E4%BD%BF%E7%94%A8telnet%E8%BF%9E%E6%8E%A5/" title="centos 使用 telnet 连接">centos 使用 telnet 连接</a><time datetime="2024-10-25T11:23:16.000Z" title="Created 2024-10-25 19:23:16">2024-10-25</time></div></div></div></div></div></div></main><footer id="footer"><div id="footer-wrap"><div class="copyright">&copy;2020 - 2024 By Yezi</div><div class="framework-info"><span>Framework </span><a target="_blank" rel="noopener" href="https://hexo.io">Hexo</a><span class="footer-separator">|</span><span>Theme </span><a target="_blank" rel="noopener" href="https://github.com/jerryc127/hexo-theme-butterfly">Butterfly</a></div></div></footer></div><div id="rightside"><div id="rightside-config-hide"><button id="readmode" type="button" title="Read Mode"><i class="fas fa-book-open"></i></button><button id="darkmode" type="button" title="Toggle Between Light And Dark Mode"><i class="fas fa-adjust"></i></button><button id="hide-aside-btn" type="button" title="Toggle between single-column and double-column"><i class="fas fa-arrows-alt-h"></i></button></div><div id="rightside-config-show"><button id="rightside_config" type="button" title="Setting"><i class="fas fa-cog fa-spin"></i></button><button class="close" id="mobile-toc-button" type="button" title="Table Of Contents"><i class="fas fa-list-ul"></i></button><a id="to_comment" href="#post-comment" title="Scroll To Comments"><i class="fas fa-comments"></i></a><button id="go-up" type="button" title="Back To Top"><span class="scroll-percent"></span><i class="fas fa-arrow-up"></i></button></div></div><div><script src="/js/utils.js"></script><script src="/js/main.js"></script><script src="https://cdn.jsdelivr.net/npm/@fancyapps/ui/dist/fancybox/fancybox.umd.min.js"></script><div class="js-pjax"><script>function loadGitalk () {
  function initGitalk () {
    var gitalk = new Gitalk(Object.assign({
      clientID: 'a1ca824ebb41b4570f94',
      clientSecret: '48b6c61bc875bb16f99774f4a271005632d6400a',
      repo: 'Yeziblog-Comments',
      owner: 'sandy9707',
      admin: ['sandy9707'],
      id: '9d8a418f0655e266fc565d27dd57bc9d',
      updateCountCallback: commentCount
    },))

    gitalk.render('gitalk-container')
  }

  if (typeof Gitalk === 'function') initGitalk()
  else {
    getCSS('https://cdn.jsdelivr.net/npm/gitalk/dist/gitalk.min.css')
    getScript('https://cdn.jsdelivr.net/npm/gitalk/dist/gitalk.min.js').then(initGitalk)
  }
}

function commentCount(n){
  let isCommentCount = document.querySelector('#post-meta .gitalk-comment-count')
  if (isCommentCount) {
    isCommentCount.innerHTML= n
  }
}

if ('Gitalk' === 'Gitalk' || !false) {
  if (false) btf.loadComment(document.getElementById('gitalk-container'), loadGitalk)
  else loadGitalk()
} else {
  function loadOtherComment () {
    loadGitalk()
  }
}</script></div><script async data-pjax src="//busuanzi.ibruce.info/busuanzi/2.3/busuanzi.pure.mini.js"></script><div id="algolia-search"><div class="search-dialog"><nav class="search-nav"><span class="search-dialog-title">Search</span><button class="search-close-button"><i class="fas fa-times"></i></button></nav><div class="search-wrap"><div id="algolia-search-input"></div><hr/><div id="algolia-search-results"><div id="algolia-hits"></div><div id="algolia-pagination"></div><div id="algolia-info"><div class="algolia-stats"></div><div class="algolia-poweredBy"></div></div></div></div></div><div id="search-mask"></div><script src="https://cdn.jsdelivr.net/npm/algoliasearch/dist/algoliasearch-lite.umd.min.js"></script><script src="https://cdn.jsdelivr.net/npm/instantsearch.js/dist/instantsearch.production.min.js"></script><script src="/js/search/algolia.js"></script></div></div></body></html>